packages/dataviews/src/components/dataviews-filters/input-widget.tsx

```diff
@@ -17,81 +17,93 @@ import type { View, NormalizedFilter, NormalizedField } from '../../types';
 import { getCurrentValue } from './utils';
 
 interface UserInputWidgetProps {
-	view: View;
-	filter: NormalizedFilter;
-	onChangeView: ( view: View ) => void;
-	fields: NormalizedField< any >[];
+    view: View;
+    filter: NormalizedFilter;
+    onChangeView: ( view: View ) => void;
+    fields: NormalizedField< any >[];
 }
 
 export default function InputWidget( {
-	filter,
-	view,
-	onChangeView,
-	fields,
+    filter,
+    view,
+    onChangeView,
+    fields,
 }: UserInputWidgetProps ) {
-	const currentFilter = view.filters?.find(
-		( f ) => f.field === filter.field
-	);
+    const currentFilter = view.filters?.find(
+        ( f ) => f.field === filter.field
+    );
 
-	const field = fields.find( ( f ) => f.id === filter.field );
-	const currentValue = getCurrentValue( filter, currentFilter );
-	const data = useMemo( () => {
-		return ( view.filters ?? [] ).reduce(
-			( acc, f ) => {
-				acc[ f.field ] = f.value;
-				return acc;
-			},
-			{} as Record< string, any >
-		);
-	}, [ view.filters ] );
+    const field = fields.find( ( f ) => f.id === filter.field );
+    const currentValue = getCurrentValue( filter, currentFilter );
 
-	const handleChange = useEvent( ( updatedData: Record< string, any > ) => {
-		if ( ! field || ! currentFilter ) {
-			return;
-		}
-		const nextValue = updatedData[ field.id ];
-		if ( fastDeepEqual( nextValue, currentValue ) ) {
-			return;
-		}
+    // Build a fake item using setValue if available, otherwise fallback to {[field.id]: value}
+    const data = useMemo( () => {
+        let item = {};
+        for ( const f of view.filters ?? [] ) {
+            const fieldDef = fields.find( fld => fld.id === f.field );
+            if (fieldDef && typeof fieldDef.setValue === 'function') {
+                item = fieldDef.setValue({ item, value: f.value });
+            } else {
+                item = { ...item, [f.field]: f.value };
+            }
+        }
+        return item;
+    }, [ view.filters, fields ] );
 
-		onChangeView( {
-			...view,
-			filters: ( view.filters ?? [] ).map( ( _filter ) =>
-				_filter.field === filter.field
-					? {
-							..._filter,
-							operator:
-								currentFilter.operator || filter.operators[ 0 ],
-							// Consider empty strings as undefined:
-							//
-							// - undefined as value means the filter is unset: the filter widget displays no value and the search returns all records
-							// - empty string as value means "search empty string": returns only the records that have an empty string as value
-							//
-							// In practice, this means the filter will not be able to find an empty string as the value.
-							value: nextValue === '' ? undefined : nextValue,
-					  }
-					: _filter
-			),
-		} );
-	} );
+    const handleChange = useEvent( ( updatedData: Record< string, any > ) => {
+        if ( !field || !currentFilter ) {
+            return;
+        }
+        // Use setValue to extract the new value for this field
+        let nextValue;
```

Comment:
At this point, all fields are normalized, and so they'll have a proper `setValue`. Additionally, if we handle this at the controls-level (see comment below), updatedData will be fine already. The only thing we'd need to do is making sure nextValue is calculated from getValue.